panther_analysis_tool/rule.py

```diff
@@ -349,16 +468,16 @@ def _run_rule(self, event: PantherEvent) -> bool:
         # defaults to True and will pass the events thru
         if self.rule_type == TYPE_SCHEDULED_RULE and not hasattr(self._module, "rule"):
             return True
-        return self._run_command(self._module.rule, event, bool)
+        return self._run_command(self._module.rule, event, bool)  # type: ignore[attr-defined]
 
     def _get_alert_context(
-        self, event: PantherEvent, use_default_on_exception: bool = True
+        self, event: PantherEvent, function_defined: bool, use_default_on_exception: bool = True
     ) -> Optional[str]:
-        if not hasattr(self._module, "alert_context"):
+        if not function_defined:
```

Comment:
Note that the `*_defined` attributes belong to `RuleResult` not the `Rule` class (which is the instance of the method here). I thought it would be more consistent to pass the function existence check result to the functions directly, rather than having them repeat the check.

Your comment made me realize that we could be passing the `RuleResult` instance to these functions and let them get any information they need from previous steps though. For example, `dedup` depends on `title`: https://github.com/panther-labs/panther_analysis_tool/pull/109/files#diff-acce689b4ddfb057c8806c0ce5a84fcf19b7163d8dee29fd1d4155d50716a415R447

Would that preferable / easier to follow?

Reply:
> Another option is that since we are checking/setting the `*_defined` in the `RuleResult` outside of these functions, we could determine it isn't set/skip the `_get_*` calls altogether and not pass this in at all?

With the exception of `_get_dedup` that is possible, but I'd have to move the default dedup string / fall-back to title outside of this method. It may be a smaller change though altogether, sounds better so I'll probably do what you suggest instead.

Reply:
> Note that the *_defined attributes belong to RuleResult not the Rule class

I see, that was where my confusion was coming in. Thanks for clarifying! I'll leave it to you @lindsey-w as to what makes the most sense here then.